### ISSUE TITLE
SIM simulation

### ISSUE THREAD

**Issue by liuxinxjtu** [charlesnchr/ML-SIM#10]
Hi @charlesnchr,  I'm very interested in your work. Why divide by w when simulating structured light frequency  in "MLSIM_datagen/SIMulator_functions.py"？
![image](https://user-images.githubusercontent.com/45114463/156482569-1902429b-2476-49e7-a287-414899462669.png)

**Comment by charlesnchr** [comment]
Hi @liuxinxjtu, thanks for your interest. 

The short answer is that `opt.k2` is not a fully formed spatial frequency due to the way that I have defined it in e.g. `GetParams()` in "SIMulator.py". Instead, this parameter represents the number of sinusoidal fringes/cycles across the entire image, and therefore to get the frequency, in units of cycles/pixel, we need to divide with the image dimension. 

**Explanation**:
Firstly, I have kept the data generation code (`MLSIM_datagen` folder) in normalised units, which means `opt.k2` does not represent a spatial frequency in physical units, but rather is defined with respect to pixel units.  

In addition to this, I prefer `opt.k2` to be the total number of cycles of the illumination pattern across the image, such that when changing the image dimensions, we get the same number of cycles (but a different frequency of course).

If you want to use physical units instead, you need to introduce a pixel size (often ~70 nm/pixel in our lab) and then you can define `opt.k2` based on an emission wavelength, for instance like this: 
```
    # Numerical aperture
    opt.NA = 1.2
    # Emission wavelength
    opt.emission = np.random.randint(500,690)
    # Pattern frequencies
    kMax = (2*opt.NA)/(opt.emission)
    factor = 0.8+0.2*np.random.rand() # Adjust for effects of aberrations
    opt.k2 = (0.4+np.random.rand()*0.2)*kMax*factor
```

My colleague @edward-n-ward prefers this approach. If you should be interested have a look at the attached file:

[SIMulator_functions_physical_units.md](https://github.com/charlesnchr/ML-SIM/files/8177421/SIMulator_functions_physical_units.md)

**Comment by liuxinxjtu** [comment]
thank you very much.

**Comment by liuxinxjtu** [comment]
Hi @charlesnchr,I have a question about the establishment of the data set. In the paper, you trained with a simulated dataset and verify the real SIM data obtained in the experiment. If the real SIM data was used as the verification set, whether the simulated data and the real SIM data obtained in the experiment are consistent with the parameters of structured light, that is, it meets the requirements of the independent and identical distribution of the data set in the neural network.

**Comment by charlesnchr** [comment]
Hi @liuxinxjtu. 
This is a central question and ultimately shows the power of the ML-SIM method. 

The parameters used in the image formation model for the simulated data does not need to exactly correspond to what is used in the lab to acquire experimental data. By generating simulated data for a wide set of parameters, the trained ML-SIM model ends up generalising beyond the configuration of a specific microscope. Fundamentally, there are some constraints to this: the dimensionality of the data should correspond (e.g. 3x5 SIM frames with resolution of 512x512 pixels) and the parameter space used for the simulated images (e.g. modulation depth, fringe orientation angle and phase shift) should contain the set of true imaging parameters (whose specific values may be unknown or at least difficult to measure accurately).

**Comment by charlesnchr** [comment]
So, to summarise, make sure that:

1. The patterns are similar (3x3 or 3x5, etc.) and the images have the same dimensions.
2. The parameter space is large enough to encapsulate the estimated true imaging parameters.

**Comment by liuxinxjtu** [comment]
thank you very much!

**Comment by charlesnchr** [comment]
Welcome. Let me know if you encounter any serious issues!
